server/container/guice/cassandra-rabbitmq-ldap-guice/src/test/java/org/apache/james/CassandraRabbitMQLdapJmapJamesServerTest.java

```diff
@@ -57,42 +59,36 @@ default void userFromLdapShouldLoginViaImapProtocol(GuiceJamesServer server) thr
     @TestInstance(TestInstance.Lifecycle.PER_CLASS)
     class WithSwift implements ContractSuite {
         @RegisterExtension
-        JamesServerExtension testExtension = baseJamesServerExtensionBuilder()
+        JamesServerExtension testExtension = baseJamesServerExtensionBuilder(objectStorage())
             .extension(new SwiftBlobStoreExtension())
-            .overrideServerModule(binder -> binder.bind(BlobStoreChoosingConfiguration.class)
-                .toInstance(BlobStoreChoosingConfiguration.objectStorage()))
             .build();
     }
 
     @Nested
     @TestInstance(TestInstance.Lifecycle.PER_CLASS)
     class WithAwsS3 implements ContractSuite {
         @RegisterExtension
-        JamesServerExtension testExtension = baseJamesServerExtensionBuilder()
+        JamesServerExtension testExtension = baseJamesServerExtensionBuilder(objectStorage())
             .extension(new AwsS3BlobStoreExtension())
-            .overrideServerModule(binder -> binder.bind(BlobStoreChoosingConfiguration.class)
-                .toInstance(BlobStoreChoosingConfiguration.objectStorage()))
             .build();
     }
 
     @Nested
     @TestInstance(TestInstance.Lifecycle.PER_CLASS)
     class WithoutSwiftOrAwsS3 implements ContractSuite {
         @RegisterExtension
-        JamesServerExtension testExtension = baseJamesServerExtensionBuilder()
-            .overrideServerModule(binder -> binder.bind(BlobStoreChoosingConfiguration.class)
-                .toInstance(BlobStoreChoosingConfiguration.cassandra()))
+        JamesServerExtension testExtension = baseJamesServerExtensionBuilder(cassandra())
             .build();
     }
 
-    JamesServerBuilder baseJamesServerExtensionBuilder() {
+    JamesServerBuilder baseJamesServerExtensionBuilder(BlobStoreConfiguration blobStoreConfiguration) {
         return new JamesServerBuilder()
             .extension(new DockerElasticSearchExtension())
             .extension(new CassandraExtension())
             .extension(new RabbitMQExtension())
             .extension(new LdapTestExtension())
             .server(configuration -> GuiceJamesServer.forConfiguration(configuration)
-                .combineWith(CassandraRabbitMQLdapJamesServerMain.MODULES)
+                .combineWith(CassandraRabbitMQLdapJamesServerMain.baseModule(blobStoreConfiguration))
```

Comment:
why BlobStoreConfiguration not part of configuration?
BTW there's a `configuration` method in `JamesServerBuilder`, we should make use of it.

Reply:
:-1:

This configuration is generic (all james server products) and putting specific stuff in it will be hard.

One step at a time?

Reply:
> -1
> 
> This configuration is generic (all james server products) and putting specific stuff in it will be hard.
> 
> One step at a time?

Yes, it's hard, of course. Meanwhile (for how long?) you propose to confuse people by having several kind of configurations when building a server for tests?

I think James is full of accidental complexity like that and we are suffering way too much because of the number of things we have to remind when writing and reading code.

Making things simpler is of course something hard but the ROI is great for us all.

We can explore the problem together if you want. Or I can try to handle it myself. But let's not surrender, please.

Reply:
Well, I'm of course not relunctant to try to reduce the complexity.

>  Meanwhile (for how long?) you propose to confuse people by having several kind of configurations when building a server for tests?

Sounds acceptable to me: the API is so far fluent enough to enable easy discovery.

IMO what you propose looks like extracting fully configuration out of guice injections, it's noble, but that's a thousand line diff that I would prefer not stacking with this thousand line diff.

Once again, our conflict burns down to the way we manage improvment. I like handling things asynchronously, in a progressive fashion and delivering things early as soon as I consider we make a positive partial step in the right direction. It don't needs to be perfect, we might not need to solve directly all the encountered problems.

My perception is that you actually want change only when it is perfect. This actually leeds to 22+ day long PR that necessarily generate suffering. This leads to simple behaviors like "loading INBOX should not read header blobs from ObjectStorage" unecessarily long to deliver. Waiting 10 month for such a behaviour actually harms our production projects.

So, if we create technical dept while writing code, when do we pay technical dept? I feel like you want to pay directly any perceived technical dept. I like paying it asynchronously. The important word is **dept** that if leveraged correctly is a powerful tool - necessary tool, any economist would tell it.

I think over the past month I proved to be paying technical dept.

So I am OK to explore a better way of configuring James. But this is loosely coupled with work here.

Actually even better: **maybe the work here enables the refactoring you have in mind**. Which is **yet another** reason to go one step at a time.

So ok to do some pair on it, but let's not delay this work.

Reply:
Please don't try to explain me economical theories, technical debt and so on, you can assume I know these things enough to have some opinions and not that I have wrong opinions because I'm not educated on these topics. It would help keeping the discussion calm and constructive.

You don't have to stack anything on this PR, nobody ask that. I would rather handle the needed refactoring in a PR than come back to this PR (that BTW would be almost trivial), 

"make the change easy then make the easy change" (Kent Beck https://twitter.com/kentbeck/status/250733358307500032 )

However, if you want to keep two different configurations, as it diverges from what we agreed upon in the ADR, we will need some documentation to explain how they interact and/or some tests to demonstrate what it does.

It's not clear to me what happens when you choose a blobstore in the configuration and you choose a blobstore in the code.

Reply:
> Please don't try to explain me economical theories, technical debt and so on, you can assume I know these things enough to have some opinions and not that I have wrong opinions because I'm not educated on these topics. It would help keeping the discussion calm and constructive.

I am more assuming a different perception.

>  as it diverges from what we agreed upon in the ADR

We said let's generate the modules based on the configuration POJO, and in test rely on the configuration pojo to do the same thing than in production.

I assumed I reached that goal here.

> It's not clear to me what happens when you choose a blobstore in the configuration and you choose a blobstore in the code.

Nothing new compared to already existing code, it is just "more visible".

Reply:
> I am more assuming a different perception.

That's funny, because in the ADR I asked for details on the solution part, and I was answered that it's simple. So I never had my details before it was merged. And now it seems that you did not perceive the solution in the same way...

> I assumed I reached that goal here.

For my part I have no idea as I still did not understand the goal.

So let's stop discussing this here, let's discuss of the goal together, eventually do some pair to check that one solution or one other is more appropriate, and update the ADR so that it's clear for anyone what is the goal of this.

Reply:
>> I am more assuming a different perception.

> That's funny, because in the ADR I asked for details on the solution part, and I was answered that it's simple. So I never had my details before it was merged. And now it seems that you did not perceive the solution in the same way...

We were not speaking of the ADR but of vague loosely related concepts like technical debt.

That's important IMO to discuss our strategy toward fighting technical debts though Github PRs might not be the best place I agree.

Regarding the ADR let's rework it then.

I suggest you write a mail explaining what you don't understand, what parts in your opinion lack details so that I can improve it.

[...]

Pair is planned.